### PR TITLE
[Sylius/Behat] Refactoring Behat extensions and pages

### PR DESCRIPTION
**Pull request by pamil**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | yes
| Deprecations? | no
| Fixed tickets | #4049
| License       | MIT
| Doc PR        | -

PageObject extension was not defensive enough, so I've done some changes to it in our extension to an extension:

  - Default `Page` class no longer extends Mink's `DocumentElement`, it has protected `Page::getDocument()` method to be used by pages only
  - There's no `Page::getPage($page)` method
  - Removed `FeatureContext`, pages should be automatically injected to the constructor

Most of our contexts should implement empty `Context` interface.

### CODE REVIEW

File: src/Sylius/Behat/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/phpspec.yml
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/.gitignore]: We never prefix files with / here. Is this gitignore needed here? I guess you should call bin/phpspec from the root dir.

File: src/Sylius/Behat/Context/Ui/CartContext.php [hunk 2]
```diff
@@ -11,48 +11,101 @@
...
+     */
+    public function myCartTaxesShouldBe($taxTotal)
+    {
+        $this->cartSummaryPage->open();
+
+        expect($this->cartSummaryPage->getTaxTotal())->toBe($taxTotal);
+    }
+
+    /**
+     * @Given my cart shipping fee should be :shippingTotal
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/CartContext.php]: Then (know it is mostly moving steps around, but let's fix it in this PR)

File: src/Sylius/Behat/Context/Ui/CartContext.php [hunk 3]
```diff
@@ -11,48 +11,101 @@
...
+     */
+    public function myCartShippingFeeShouldBe($shippingTotal)
+    {
+        expect($this->cartSummaryPage->getShippingTotal())->toBe($shippingTotal);
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Ui/CartContext.php]: Me likes :D

File: src/Sylius/Behat/.gitignore (hunk 1 above)
Comment (pamil) [inline src/Sylius/Behat/.gitignore]: This way I can call `../../../bin/phpspec run` from that directory and have only that specifications runned, what results in a faster feedback cycle.

IMO ignoring absolute paths is better than any file named `phpspec.yml`, I only want to ignore `src/Sylius/Behat/phpspec.yml` but not `src/Sylius/Behat/foo/bar/phpspec.yml`.

File: src/Sylius/Behat/Context/Ui/CartContext.php (hunk 2 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/CartContext.php]: :+1:

File: src/Sylius/Behat/.gitignore (hunk 1 above)
Comment (pjedrzejewski) [inline src/Sylius/Behat/.gitignore]: bin/phpspec run src/Sylius/Behat/spec?

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 4]
```diff
@@ -74,6 +74,7 @@ public function __construct(
     /**
      * @Transform /^product "([^"]+)"$/
      * @Transform /^"([^"]+)" product$/
+     * @Transform :product
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: `:productName`?

File: src/Sylius/Behat/Page/Product/ProductShowPage.php [hunk 5]
```diff
@@ -20,13 +19,13 @@
 class ProductShowPage extends SymfonyPage
 {
     /**
-     * @param ProductInterface $product
+     * @param array $urlParameters
      *
      * @return ProductShowPage
      */
-    public function openSpecificProductPage(ProductInterface $product)
+    public function open(array $urlParameters = [])
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Page/Product/ProductShowPage.php]: What was wrong with previous signature?

File: src/Sylius/Behat/PageObjectExtension/PageObject/Page.php [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    /**
+     * @param string $statusCode
+     *
+     * @return bool
+     */
+    protected function isErrorResponse($statusCode)
+    {
+        return in_array(substr($statusCode, 0, 1), ['4', '5'], true);
```
Comment (michalmarcinkowski) [inline src/Sylius/Behat/PageObjectExtension/PageObject/Page.php]: The `$statusCode` is `int` so we can check `400 <= $statusCode && $statusCode < 600`

File: src/Sylius/Behat/Context/Setup/ProductContext.php (hunk 4 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: We should use `:productName` in step definitions if we want to get product name. E.g.
```gherkin
I create :productName # I want to get product name as a string
I add :product to the cart # I want to get product entity
```

File: src/Sylius/Behat/Page/Product/ProductShowPage.php (hunk 5 above)
Comment (pamil) [inline src/Sylius/Behat/Page/Product/ProductShowPage.php]: The `open()` wasn't opening a page and it was the only exception from the rule. If we manage to specify pages as a services, every page will be able to define their own opening methods.

File: src/Sylius/Behat/Context/Setup/ProductContext.php (hunk 4 above)
Comment (michalmarcinkowski) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: :+1:

**Comment by pjedrzejewski** [comment]
Awesome work, thanks Kamil!

**Pull request merged**